File: android/app/src/main/java/com/izzatismail/midtrim/domain/usecase/ImportVideosUseCase.kt
package com.izzatismail.midtrim.domain.usecase

import com.izzatismail.midtrim.domain.entity.VideoMetadata
import com.izzatismail.midtrim.domain.error.ImportVideoError
import com.izzatismail.midtrim.domain.repository.VideoMetadataService

class ImportVideosUseCase(
    private val metadataService: VideoMetadataService,
    private val fetchEntitlementStatus: FetchEntitlementStatusUseCase
) {
    private val maxVideos = 10
    private val maxVideosPaid = 20

    suspend fun execute(videoUris: List<String>): List<VideoMetadata> {
        if (videoUris.isEmpty()) throw ImportVideoError.EmptySelection

        val cap = if (fetchEntitlementStatus.isPaidUser) maxVideosPaid else maxVideos
        if (videoUris.size > cap) throw ImportVideoError.ExceedsTierCap(cap, videoUris.size)

        return videoUris.map { uri ->
            metadataService.fetchMetadata(uri)
        }
    }
}

File: ios/MidTrim/MidTrim/Domain/UseCases/ImportVideosUseCase.swift
import Foundation

struct ImportVideosUseCase {
    private let metadataService: VideoMetadataServiceProtocol
    private let entitlementCache: FetchEntitlementStatusUseCase

    init(metadataService: VideoMetadataServiceProtocol, entitlementCache: FetchEntitlementStatusUseCase) {
        self.metadataService = metadataService
        self.entitlementCache = entitlementCache
    }

    private var maxVideos: Int { 10 }
    private var maxVideosPaid: Int { 20 }

    func execute(videoURIs: [String]) async throws -> [VideoMetadata] {
        guard !videoURIs.isEmpty else {
            throw ImportVideoError.emptySelection
        }

        let isPaid = await entitlementCache.isPaidUser
        let cap = isPaid ? maxVideosPaid : maxVideos

        guard videoURIs.count <= cap else {
            throw ImportVideoError.exceedsTierCap(max: cap, attempted: videoURIs.count)
        }

        var metadataList: [VideoMetadata] = []
        for uri in videoURIs {
            let metadata = try await metadataService.fetchMetadata(for: uri)
            metadataList.append(metadata)
        }
        return metadataList
    }
}
